internal/locate/sorted_btree.go

```diff
@@ -79,12 +80,22 @@ func (s *SortedRegions) SearchByKey(key []byte, isEndKey bool) (r *Region) {
 }
 
 // AscendGreaterOrEqual returns all items that are greater than or equal to the key.
+// It is the caller's responsibility to make sure that startKey is a node in the B-tree, otherwise, the startKey will not be included in the return regions.
 func (s *SortedRegions) AscendGreaterOrEqual(startKey, endKey []byte, limit int) (regions []*Region) {
+	now := time.Now().Unix()
+	lastStartKey := startKey
 	s.b.AscendGreaterOrEqual(newBtreeSearchItem(startKey), func(item *btreeItem) bool {
 		region := item.cachedRegion
 		if len(endKey) > 0 && bytes.Compare(region.StartKey(), endKey) >= 0 {
 			return false
 		}
+		if !region.checkRegionCacheTTL(now) {
+			return false
+		}
+		if len(lastStartKey) > 0 && !region.Contains(lastStartKey) { // uncached hole
```

Comment:
when will `len(lastStartKey) == 0`?

Reply:
Some test cases have the key, for -inf to +inf.

Reply:
The case `len(lastStartKey) == 0` should never happen in current usage(start key is assigned [here](https://github.com/tikv/client-go/blob/9e5fff01aef323cbb5b38699e65d64278a6ebb7c/internal/locate/region_cache.go#L1259), and it's never empty if the code runs here).

When `len(lastStartKey) == 0` and the first region is not start from -inf, we should stop scan regions, see the following test.

```go
func (s *testRegionRequestToSingleStoreSuite) TestRegionCacheStartNonEmpty() {
	_ = s.cache.refreshRegionIndex(s.bo)
	r, _ := s.cache.scanRegionsFromCache(s.bo, []byte{}, nil, 10)
	s.Equal(len(r), 1)

	region, _ := s.cache.LocateRegionByID(s.bo, s.region)
	v2 := region.Region.confVer + 1
	r2 := metapb.Region{Id: region.Region.id, RegionEpoch: &metapb.RegionEpoch{Version: region.Region.ver, ConfVer: v2}, StartKey: []byte{1}}
	st := newUninitializedStore(s.store)

	s.cache.mu.Lock()
	s.cache.mu.sorted.Clear()
	s.cache.mu.Unlock()
	// region cache after clear: []

	s.cache.insertRegionToCache(&Region{meta: &r2, store: unsafe.Pointer(st), ttl: nextTTLWithoutJitter(time.Now().Unix())}, true, true)
	// region cache after insert: [[1, +inf)]

	r, _ = s.cache.scanRegionsFromCache(s.bo, []byte{}, nil, 10)
	s.Equal(len(r), 0) // fail in current implementation
}
```

Reply:
I add the test `TestRegionCacheStartNonEmpty` and remove the `len(lastStartKey) == 0` check.  Now the code stop scan regions when `len(lastStartKey) == 0` and the first region is not start from -inf.   PTAL again.